scanpy/preprocessing/normalization.py

```diff
@@ -0,0 +1,109 @@
+import numpy as np
+from scipy.sparse import issparse
+from sklearn.utils import sparsefuncs
+from .. import logging as logg
+from ..utils import doc_params
+from .docs import doc_norm_bulk, doc_norm_quant, doc_ret, doc_ex_quant, doc_ex_total
+
+def _normalize_data(X, counts, after=None, copy=False):
+    X = X.copy() if copy else X
+    after = np.median(counts[counts>0]) if after is None else after
+    counts += (counts == 0)
+    counts /= after
+    if issparse(X):
+        X = sparsefuncs.inplace_row_scale(X, 1/counts)
+    else:
+        X /= counts[:, None]
+    return X if copy else None
+
+@doc_params(norm_bulk=doc_norm_bulk, norm_quant=doc_norm_quant, ret=doc_ret, ex_quant=doc_ex_quant)
+def normalize_quantile(data, cell_sum_after=None, quantile=1, key_n_counts=None,
+                       inplace=True, layers=[], layer_norm=None):
+    """\
+    {norm_bulk}
+    {norm_quant}
+
+    {ret}
+
+    {ex_quant}
+    """
+    if quantile < 0 or quantile > 1:
+        raise ValueError('Choose quantile between 0 and 1.')
+
+    X = data.X
+    gene_subset = None
+    if not inplace:
+    # not recarray because need to support sparse
+        dat = {}
+
+    if quantile < 1:
+        logg.msg('normalizing by count per cell for \
+                  genes that make up less than quantile * total count per cell', r=True)
+        X = data.X
+
+        counts_per_cell = X.sum(1)
+        counts_per_cell = np.ravel(counts_per_cell)
+
+        gene_subset = (X>counts_per_cell[:, None]*quantile).sum(0)
+        gene_subset = (np.ravel(gene_subset) == 0)
+    else:
+        logg.msg('normalizing by total count per cell', r=True)
+
+    X = X if gene_subset is None else data[:, gene_subset].X
+    counts_per_cell = X.sum(1)
+    #get rid of data view
+    counts_per_cell = np.ravel(counts_per_cell).copy()
+    del X
+    del gene_subset
```

Comment:
No necessity for deleting the objects. It will not free memory anyways unless you call a garbage collector... Simply remove...

Reply:
Yes, this only deletes the references. The logic is to delete references which i don't use later. I don't know when garbage collection happen precisely but i believe it may happen during the execution of the function and then it will free the memory (so it is not dragged through the whole function). If it doesn't happen, then it is no worse than not using del...

But i delete it of course, if you think that it clutters the function.

Reply:
I’d rather recommend to use helper functions for the main steps of the long main function.

1. less spaghetti, the main function is basically just a list of steps!
2. get GC without having to sprinkle “del”
3. you can’t miss deleting a variable you don’t need anymore